storage/fanout.go

```diff
@@ -481,59 +554,75 @@ func (c *mergeSeriesSet) Err() error {
 	return nil
 }
 
-type seriesSetHeap []SeriesSet
+type genericSeriesSetHeap []genericSeriesSet
 
-func (h seriesSetHeap) Len() int      { return len(h) }
-func (h seriesSetHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }
+func (h genericSeriesSetHeap) Len() int      { return len(h) }
+func (h genericSeriesSetHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }
 
-func (h seriesSetHeap) Less(i, j int) bool {
+func (h genericSeriesSetHeap) Less(i, j int) bool {
 	a, b := h[i].At().Labels(), h[j].At().Labels()
 	return labels.Compare(a, b) < 0
 }
 
-func (h *seriesSetHeap) Push(x interface{}) {
-	*h = append(*h, x.(SeriesSet))
+func (h *genericSeriesSetHeap) Push(x interface{}) {
+	*h = append(*h, x.(genericSeriesSet))
 }
 
-func (h *seriesSetHeap) Pop() interface{} {
+func (h *genericSeriesSetHeap) Pop() interface{} {
 	old := *h
 	n := len(old)
 	x := old[n-1]
 	*h = old[0 : n-1]
 	return x
 }
 
-type mergeSeries struct {
+// ChainedSeriesMerge returns single series from many same series by chaining samples together.
+// In case of the timestamp overlap, first overlapped sample is kept, rest are dropped.
+// We expect the same labels for each given series.
+// TODO(bwplotka): This has the same logic as tsdb.verticalChainedSeries. Remove this in favor of ChainedSeriesMerge in next PRs.
```

Comment:
So the general plan is to remove the pluggability of the merge function, and only have vertical merging everywhere?

Reply:
Correct. Commented here: https://github.com/prometheus/prometheus/pull/7005#issuecomment-602475951